### PR TITLE
New APIs to reject unknown values.

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

**Comment by swankjesse** [comment]
I'm a little uncertain about where this API is slotted in.

File: moshi/src/main/java/com/squareup/moshi/JsonAdapter.java [hunk 1]
```diff
@@ -103,6 +103,29 @@ public final String toJson(T value) {
     };
   }
 
+  /**
+   * Returns a JSON adapter equal to this, but that throws a {@link JsonDataException} when {@link
```
Comment (JakeWharton) [inline moshi/src/main/java/com/squareup/moshi/JsonAdapter.java]: nit: linkplain

File: moshi/src/main/java/com/squareup/moshi/JsonReader.java [hunk 2]
```diff
@@ -214,6 +214,9 @@
   /** True to accept non-spec compliant JSON */
   private boolean lenient = false;
 
+  /** True to throw a {@link JsonDataException} on any attempt to call {@link #skipValue()}. */
+  private boolean skipForbidden = false;
```
Comment (JakeWharton) [inline moshi/src/main/java/com/squareup/moshi/JsonReader.java]: When I first read this, I thought it was a boolean to control skipping something classified as a forbidden value. Not forbidding the act of skipping values.

Comment (JakeWharton) [inline moshi/src/main/java/com/squareup/moshi/JsonReader.java]: Should this be `skipAllowed`? I usually find booleans that assert in the positive direction to be easier to reason about.

Think fast: what does `setSkipForbidden(false)` do?

**Comment by JakeWharton** [comment]
@SimonVT this change is for you!

File: moshi/src/test/java/com/squareup/moshi/JsonReaderTest.java [hunk 3]
```diff
@@ -191,6 +190,42 @@
...
+    reader.setSkipForbidden(true);
+    reader.beginObject();
+    assertThat(reader.nextName()).isEqualTo("a");
+    try {
+      reader.skipValue();
+      fail();
+    } catch (JsonDataException expected) {
+      assertThat(expected).hasMessage("Cannot skip unexpected NUMBER at $.a");
+    }
+    // Confirm that the reader is left in a consistent state after the exception.
```
Comment (JakeWharton) [inline moshi/src/test/java/com/squareup/moshi/JsonReaderTest.java]: How considerate!

**Comment by JakeWharton** [comment]
LGTM

File: moshi/src/main/java/com/squareup/moshi/JsonReader.java (hunk 2 above)
Comment (swankjesse) [inline moshi/src/main/java/com/squareup/moshi/JsonReader.java]: How do you feel about this: `JsonReader.setFailOnUnknown(true)` ?

File: moshi/src/test/java/com/squareup/moshi/JsonReaderTest.java (hunk 3 above)
Comment (swankjesse) [inline moshi/src/test/java/com/squareup/moshi/JsonReaderTest.java]: Yep. Though this behavior might be the first to go if it harms performance.

File: moshi/src/main/java/com/squareup/moshi/JsonReader.java (hunk 2 above)
Comment (swankjesse) [inline moshi/src/main/java/com/squareup/moshi/JsonReader.java]: ... that's what I'm going with.

Comment (JakeWharton) [inline moshi/src/main/java/com/squareup/moshi/JsonReader.java]: :+1:

**Comment by SimonVT** [comment]
@JakeWharton haha, awesome! But now I have to look into using moshi...

File: moshi/src/main/java/com/squareup/moshi/JsonAdapter.java [hunk 4]
```diff
@@ -103,6 +103,30 @@ public final String toJson(T value) {
...
   }
 
+  /**
+   * Returns a JSON adapter equal to this, but that throws a {@link JsonDataException} when
+   * {@linkplain JsonReader#setFailOnUnknown(boolean) unknown values} are encountered. This
+   * constraint applies to both the top-level message handled by this type adapter as well as to
+   * nested messages.
+   */
+  public final JsonAdapter<T> failOnUnknown() {
+    final JsonAdapter<T> delegate = this;
```
Comment (inder123) [inline moshi/src/main/java/com/squareup/moshi/JsonAdapter.java]: Instead of failing, you should consider using a callback for the programmer to handle it.

Comment (swankjesse) [inline moshi/src/main/java/com/squareup/moshi/JsonAdapter.java]: Interesting idea. So the parser could log a warning but not crash?

Comment (inder123) [inline moshi/src/main/java/com/squareup/moshi/JsonAdapter.java]: How do you log without passing an explicit logger to Moshi?

Just allow the programmer to set a callback to receive such events and handle them in some way.

**Pull request merged**